### PR TITLE
Adding more grover layers

### PR DESCRIPTION
**Pull request by gusty1g**
# Pull Request Template

## Description

I am breaking down #3272 into two pull requests. Here is the part one, containing some layers in the grover pretrain model.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/grover_layers.py [hunk 1]
```diff
@@ -1,16 +1,263 @@
...
+        Dynamic message passing depth for use in MPNEncoder
+    undirected: bool
+        The message passing is undirected or not
+    num_mt_block: int
+        the number of message passing blocks.
+    num_head: int
+        the number of attention heads.
+    atom_embedding_output_type: str
+        the type of output aggregation after message passing.
+                                            atom_messages:      True                      False
```
Comment (rbharath) [inline deepchem/models/torch_models/grover_layers.py:37]: Is the formatting here messed up? It looks strange

File: deepchem/models/torch_models/grover_layers.py [hunk 2]
```diff
@@ -1,16 +1,263 @@
...
+            atom_emb_output_type=atom_embedding_output_type,
+            bias=bias,
+            res_connection=res_connection)
+
+    def forward(self, graph_batch: List[torch.Tensor]):
+        """Forward function
+
+        Parameters
+        ----------
+        graph_batch: List[torch.Tensor]
```
Comment (rbharath) [inline deepchem/models/torch_models/grover_layers.py:79]: Can you add a `Returns` field here?

File: deepchem/models/torch_models/grover_layers.py [hunk 3]
```diff
@@ -1,16 +1,263 @@
...
+                "bond_from_atom": output[0][1],
+                "atom_from_bond": output[1][0],
+                "bond_from_bond": output[1][1]
+            }
+
+
+class GroverBondVocabPredictor(nn.Module):
+    """Layer for learning contextual information for bonds.
+
+    The layer is used in `Grover <https://drug.ai.tencent.com/publications/GROVER.pdf>`_
```
Comment (rbharath) [inline deepchem/models/torch_models/grover_layers.py:109]: Can you add a `References` section and move the link there?

File: deepchem/models/torch_models/grover_layers.py [hunk 4]
```diff
@@ -1,16 +1,263 @@
...
+    architecture to learn contextual information of a bond by predicting
+    the context of a bond from the bond embedding in a multi-class classification setting.
+    The contextual information of a bond are encoded as strings (ex: '(DOUBLE-STEREONONE-NONE)_C-(SINGLE-STEREONONE-NONE)2').
```
Comment (rbharath) [inline deepchem/models/torch_models/grover_layers.py:112]: Can you add a usage example?

File: deepchem/models/torch_models/grover_layers.py [hunk 5]
```diff
@@ -1,16 +1,263 @@
...
+        logits = self.linear(embeddings[ids1]) + self.linear_rev(
+            embeddings[ids2])
+        return self.logsoftmax(logits)
+
+
+class GroverAtomVocabPredictor(nn.Module):
+    """Grover Atom Vocabulary Prediction Module.
+
+    The GroverAtomVocabPredictor module is used for predicting atom-vocabulary
+    for the self-supervision task in `Grover. <https://drug.ai.tencent.com/publications/GROVER.pdf>`_
```
Comment (rbharath) [inline deepchem/models/torch_models/grover_layers.py:156]: Same comments as before. Can you add `References` and can you add a usage example?

File: deepchem/models/torch_models/grover_layers.py [hunk 6]
```diff
@@ -1,16 +1,263 @@
...
+            the prediction for each atom of shape (num_bond, vocab_size)
+        """
+        return self.logsoftmax(self.linear(embeddings))
+
+
+class GroverFunctionalGroupPrediction(nn.Module):
+    """The functional group prediction task for self-supervised learning.
+
+    Molecules have functional groups in them. This module is used for predicting
+    the functional group and the problem is formulated as an multi-label classification problem.
```
Comment (rbharath) [inline deepchem/models/torch_models/grover_layers.py:196]: Can you add a usage example?

**Review by rbharath — COMMENTED** [review]
Mostly minor comments here requesting improvements to the docs

File: deepchem/models/torch_models/grover_layers.py (hunk 4 above)
Comment (gusty1g) [reply deepchem/models/torch_models/grover_layers.py:110]: Added example

File: deepchem/models/torch_models/grover_layers.py (hunk 1 above)
Comment (gusty1g) [reply deepchem/models/torch_models/grover_layers.py]: Fixed it

File: deepchem/models/torch_models/grover_layers.py (hunk 2 above)
Comment (gusty1g) [reply deepchem/models/torch_models/grover_layers.py:73]: added returns field

File: deepchem/models/torch_models/grover_layers.py (hunk 3 above)
Comment (gusty1g) [reply deepchem/models/torch_models/grover_layers.py]: added reference section

File: deepchem/models/torch_models/grover_layers.py [hunk 7]
```diff
@@ -1,16 +1,310 @@
...
+        """Initializes GroverBondVocabPredictor
+
+        Parameters
+        ----------
+        vocab_size: int
+            Size of vocabulary, used for number of classes in prediction.
+        in_features: int, default: 128
+            Input feature size of bond embeddings.
+
+        Example
```
Comment (rbharath) [inline deepchem/models/torch_models/grover_layers.py:127]: This should be in the main docstring above and not in the init docstring

**Review by rbharath — COMMENTED** [review]
One minor fix needed on the docstring and we should be good to go

**Review by rbharath — APPROVED** [review]
LGTM. Feel free to merge in once CI looks good

**Pull request merged**